File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/FindInsteadOfFirstOrDefault.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

public class FindInsteadOfFirstOrDefault
{
    public class Dummy
    {
        public object FirstOrDefault() => null;
        public object FirstOrDefault(Func<int, bool> predicate) => null;
    }

    public class AnotherDummy
    {
        public object FirstOrDefault => null;
    }

    public void UnrelatedType(Dummy dummy, AnotherDummy anotherDummy)
    {
        dummy.FirstOrDefault(); // Compliant
        dummy.FirstOrDefault(x => true); // Compliant

        _ = anotherDummy.FirstOrDefault;

        // Nullable
        dummy?.FirstOrDefault(); // Compliant
        dummy?.FirstOrDefault(x => true); // Compliant

        _ = anotherDummy?.FirstOrDefault;
    }

    public class MyList : List<int>
    {
        public MyList Fluent() => this;
    }

    public class HiddenList : List<int>
    {
        public int? FirstOrDefault(Func<int, bool> predicate) => null;
    }

    public static class HelperClass
    {
        public static List<int> DoWorkReturnGroup() => null;
        public static void DoWorkMethodGroup<T>(Func<Func<T, bool>, T> firstOrDefault) { }

        public static bool FilterMethod(int nb) => true;
    }

    public bool FilterMethod(int nb) => true;

    public void ListBasic(List<int> data)
    {
        data.FirstOrDefault(x => true); // Noncompliant {{"Find" method should be used instead of the "FirstOrDefault" extension method.}}
        //   ^^^^^^^^^^^^^^
        data.Find(x => true); // Compliant

        data.FirstOrDefault(); // Compliant
        data.FirstOrDefault(HelperClass.FilterMethod); // Noncompliant
        //   ^^^^^^^^^^^^^^
        data.FirstOrDefault(FilterMethod); // Noncompliant
        //   ^^^^^^^^^^^^^^
    }

    public void ThroughLinq(List<int> data)
    {
        data.Select(x => x * 2).ToList().FirstOrDefault(x => true); // Noncompliant
        //                               ^^^^^^^^^^^^^^
        data.Select(x => x * 2).ToList().Find(x => true); // Compliant
    }

    public void ThroughFunction()
    {
        HelperClass.DoWorkReturnGroup().FirstOrDefault(x => true); // Noncompliant
        //                              ^^^^^^^^^^^^^^
        HelperClass.DoWorkReturnGroup().Find(x => true); // Compliant
    }

    public void ThroughLambda(Func<List<int>> lambda)
    {
        lambda().FirstOrDefault(x => true); // Noncompliant
        //       ^^^^^^^^^^^^^^
        lambda().Find(x => true); // Compliant
    }

    public void WithinALambda()
    {
        _ = new Func<List<int>, int>(list => list.FirstOrDefault(x => true)); // Noncompliant
        //                                        ^^^^^^^^^^^^^^
    }

    public void AsMethodGroup(List<int> data)
    {
        HelperClass.DoWorkMethodGroup<int>(data.FirstOrDefault); // FN, this raise as a SimpleAccessMemberExpression
    }

    public void Miscellaneous(List<int> data)
    {
        _ = nameof(Enumerable.FirstOrDefault); // Compliant
    }

    public static void FirstOrdefaultNotHidden(MyList data)
    {
        data.FirstOrDefault(x => true); // Noncompliant
        //   ^^^^^^^^^^^^^^
        data.Find(x => true); // Compliant
    }

    public static void FirstOrdefaultHidden(HiddenList data)
    {
        data.FirstOrDefault(x => true); // Compliant
        data.Find(x => true); // Compliant
    }

    public static void Nullable(List<int> data = null)
    {
        data?.FirstOrDefault(x => true); // Noncompliant
        //    ^^^^^^^^^^^^^^
        data?.Find(x => true); // Compliant
    }

    public static void SpecialPattern(List<int> data)
    {
        (true ? data : data).FirstOrDefault(x => true); // Noncompliant
        //                   ^^^^^^^^^^^^^^
        (data ?? data).FirstOrDefault(x => true); // Noncompliant
        //             ^^^^^^^^^^^^^^
        (data ?? (true ? data : data)).FirstOrDefault(x => true); // Noncompliant
        //                             ^^^^^^^^^^^^^^
    }

    public static void FluentNullable(MyList data = null)
    {
        data.Fluent().Fluent().Fluent().Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                       ^^^^^^^^^^^^^^
        data.Fluent().Fluent().Fluent().Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                        ^^^^^^^^^^^^^^
        data.Fluent().Fluent().Fluent()?.Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                        ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent().Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                        ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent().Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                        ^^^^^^^^^^^^^^
        data.Fluent().Fluent().Fluent()?.Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                         ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent().Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                         ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent()?.Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                         ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent().Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                         ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent()?.Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                         ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent().Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                         ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent()?.Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent()?.Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent().Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent()?.Fluent().FirstOrDefault(x => true); // Noncompliant
        //                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent()?.Fluent()?.FirstOrDefault(x => true); // Noncompliant
        //                                           ^^^^^^^^^^^^^^
    }
}

public class ExpressionTree
{
    public void InExpressionTree()
    {
        Expression<Func<List<int>, int>> firstThree = list => list.FirstOrDefault(el => el == 3); // Compliant (IsInExpressionTree)
    }

    public List<string> Repro_7964(IQueryable<List<string>> values) =>                 // See https://github.com/SonarSource/sonar-dotnet/issues/7964
        values.FirstOrDefault(p => p.FirstOrDefault(x => x.Equals("a")) != null); // Compliant - usage in expression tree
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/FindInsteadOfFirstOrDefault.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System

Module HelperClass
    Function DoWorkReturnGroup() As List(Of Integer)
        Return Nothing
    End Function

    Sub DoWorkMethodGroup(Of T)(firstOrDefault As Func(Of Func(Of T, Boolean), T))
    End Sub

    Function FilterMethod(nb As Integer) As Boolean
        Return true
    End Function
End Module

Public Class FindInsteadOfFirstOrDefault
    Public Class Dummy
        Public Function FirstOrDefault() As Object
            Return Nothing
        End Function

        Public Function FirstOrDefault(predicate As Func(Of Integer, Boolean)) As Object
            Return Nothing
        End Function
    End Class

    Public Class AnotherDummy
        Public ReadOnly Property FirstOrDefault As Object
            Get
                Return Nothing
            End Get
        End Property
    End Class

    Public Sub UnrelatedType(ByVal dummy As Dummy, ByVal anotherDummy As AnotherDummy)
        Dim unused = dummy.FirstOrDefault() ' Compliant
        unused = dummy.FirstOrDefault(Function(x) True) ' Compliant

        unused = anotherDummy.FirstOrDefault
    End Sub

    Public Class MyList
        Inherits List(Of Integer)

        Public Function Fluent() As MyList
            Return Me
        End Function
    End Class

    Public Class HiddenList
        Inherits List(Of Integer)

        Public Function FirstOrDefault(predicate As Func(Of Integer, Boolean)) As Integer?
            Return Nothing
        End Function
    End Class

    Public Function FilterMethod(nb As Integer) As Boolean
        Return True
    End Function

    Public Sub ListBasic(ByVal data As List(Of Integer))
        Dim unused = data.FirstOrDefault(Function(x) True) ' Noncompliant
        '                 ^^^^^^^^^^^^^^
        unused = data.Find(Function(x) True) ' Compliant

        unused = data.FirstOrDefault() ' Compliant
        unused = data.FirstOrDefault(AddressOf HelperClass.FilterMethod) ' Noncompliant
        '             ^^^^^^^^^^^^^^
        unused = data.FirstOrDefault(AddressOf FilterMethod) ' Noncompliant
        '             ^^^^^^^^^^^^^^
    End Sub

    Public Sub ThroughLinq(data As List(Of Integer))
        data.[Select](Function(x) x * 2).ToList().FirstOrDefault(Function(x) True) ' Noncompliant {{"Find" method should be used instead of the "FirstOrDefault" extension method.}}
        '                                         ^^^^^^^^^^^^^^
        data.[Select](Function(x) x * 2).ToList().Find(Function(x) True) ' Compliant
    End Sub

    Public Sub ThroughFunction()
        Dim unused = HelperClass.DoWorkReturnGroup().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                            ^^^^^^^^^^^^^^
        unused = HelperClass.DoWorkReturnGroup().Find(Function(x) True) ' Compliant
    End Sub

    Public Sub ThroughLambda(lambda As Func(Of List(Of Integer)))
        Dim unused = lambda().FirstOrDefault(Function(x) True) ' Noncompliant
        '                     ^^^^^^^^^^^^^^
        unused = lambda().Find(Function(x) True) ' Compliant
    End Sub

    Public Sub WithinALambda()
        Dim unused = New Func(Of List(Of Integer), Integer)(Function(list) list.FirstOrDefault(Function(x) True)) ' Noncompliant
        '                                                                       ^^^^^^^^^^^^^^
    End Sub

    Public Sub AsMethodGroup(data As List(Of Integer))
        HelperClass.DoWorkMethodGroup(Of Integer)(AddressOf data.FirstOrDefault) ' FN
    End Sub

    Public Sub SpecialPattern(data As List(Of Integer))
        Dim unused = (If(True, data, data)).FirstOrDefault(Function(x) True) ' Noncompliant
        '                                   ^^^^^^^^^^^^^^
        unused = (If(data, data)).FirstOrDefault(Function(x) True) ' Noncompliant
        '                         ^^^^^^^^^^^^^^
        unused = (If(data, (If(True, data, data)))).FirstOrDefault(Function(x) True) ' Noncompliant
        '                                           ^^^^^^^^^^^^^^
    End Sub

    Public Sub Fluent(data As MyList)
        data.Fluent().Fluent().Fluent().Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                        ^^^^^^^^^^^^^^
        data.Fluent().Fluent().Fluent().Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                         ^^^^^^^^^^^^^^
        data.Fluent().Fluent().Fluent()?.Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                         ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent().Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                         ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent().Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                         ^^^^^^^^^^^^^^
        data.Fluent().Fluent().Fluent()?.Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                          ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent().Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                          ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent()?.Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent().Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent()?.Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                          ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent().Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                          ^^^^^^^^^^^^^^
        data.Fluent().Fluent()?.Fluent()?.Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                           ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent().Fluent()?.Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                           ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent().Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                           ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent()?.Fluent().FirstOrDefault(Function(x) True) ' Noncompliant
        '                                           ^^^^^^^^^^^^^^
        data.Fluent()?.Fluent()?.Fluent()?.Fluent()?.FirstOrDefault(Function(x) True) ' Noncompliant
        '                                            ^^^^^^^^^^^^^^
    End Sub
End Class
